File: enabled-product-missing-from-storefront/python/diagnose_missing_product.py
"""Diagnose a Magento 2 product that shows Enabled but is missing from the storefront.

Status Enabled is only one of several conditions Magento checks. Visibility has
to include catalog or search, the product has to carry the storefront's
website_id in its website assignment (a REST create/update that omits
extension_attributes.website_ids can silently drop or fail to set this, per
magento2 GitHub issues #8173, #10495, #11324), and it has to link to at least
one active category. Even when all three agree, a stale or invalid indexer
(catalog_category_product, catalog_product_index, catalogsearch_fulltext) or a
missed cron run can still hide the product, and that can only be fixed with
bin/magento indexer:reindex, not over REST. This reports by default. Run on a
schedule or on demand. Safe to run again and again.

Guide: https://www.allanninal.dev/magento/enabled-product-missing-from-storefront/
"""
import os
import logging
import requests

logging.basicConfig(level=logging.INFO, format="%(asctime)s %(levelname)s %(message)s")
log = logging.getLogger("diagnose_missing_product")

MAGENTO_URL = os.environ["MAGENTO_URL"].rstrip("/")
TOKEN = os.environ["MAGENTO_ADMIN_TOKEN"]
TARGET_WEBSITE_ID = int(os.environ.get("TARGET_WEBSITE_ID", "1"))
DRY_RUN = os.environ.get("DRY_RUN", "true").lower() == "true"

NOT_VISIBLE_INDIVIDUALLY = 1
SUSPECT_INDEXERS = ["catalog_category_product", "catalog_product_index", "catalogsearch_fulltext"]


def magento_get(path, params=None):
    r = requests.get(
        f"{MAGENTO_URL}/rest/V1{path}",
        params=params or {},
        headers={"Authorization": f"Bearer {TOKEN}"},
        timeout=30,
    )
    r.raise_for_status()
    return r.json()


def magento_put(path, body):
    r = requests.put(
        f"{MAGENTO_URL}/rest/V1{path}",
        json=body,
        headers={"Authorization": f"Bearer {TOKEN}", "Content-Type": "application/json"},
        timeout=30,
    )
    r.raise_for_status()
    return r.json()


def decide_storefront_eligibility(product, categories, target_website_id):
    """Pure decision logic, no I/O.

    product: {status: 1|2, visibility: 1|2|3|4, websiteIds: [int], categoryIds: [int]}
    categories: [{id: int, isActive: bool}]
    target_website_id: int
    Returns {"eligible": bool, "reasons": [str]}.

    eligible is True only if status == 1 AND visibility != 1 (Not Visible
    Individually) AND target_website_id is in websiteIds AND at least one
    linked category is active. Otherwise reasons lists every failing
    condition, so the caller can tell "should be eligible per data but still
    missing from storefront" (stale index/cron) apart from "genuinely
    ineligible per data".
    """
    reasons = []

    if product["status"] != 1:
        reasons.append("disabled")

    if product["visibility"] == NOT_VISIBLE_INDIVIDUALLY:
        reasons.append("not_visible_individually")

    if target_website_id not in product["websiteIds"]:
        reasons.append("website_not_assigned")

    active_ids = {c["id"] for c in categories if c["isActive"]}
    if not any(cid in active_ids for cid in product["categoryIds"]):
        reasons.append("no_active_category")

    return {"eligible": len(reasons) == 0, "reasons": reasons}


def find_product_by_sku(sku):
    params = {
        "searchCriteria[filterGroups][0][filters][0][field]": "sku",
        "searchCriteria[filterGroups][0][filters][0][value]": sku,
        "searchCriteria[filterGroups][0][filters][0][conditionType]": "eq",
    }
    items = magento_get("/products", params)["items"]
    return items[0] if items else None


def visibility_of(product):
    for attr in product.get("custom_attributes", []):
        if attr["attribute_code"] == "visibility":
            return int(attr["value"])
    return None


def category_ids_of(product):
    for attr in product.get("custom_attributes", []):
        if attr["attribute_code"] == "category_ids":
            return [int(c) for c in attr["value"]]
    return []


def product_website_ids(sku):
    return magento_get(f"/products/{sku}/websites")


def fetch_categories(category_ids):
    result = []
    for cid in category_ids:
        data = magento_get(f"/categories/{cid}")
        result.append({"id": cid, "isActive": bool(data.get("is_active"))})
    return result


def storefront_has_product(storefront_url):
    r = requests.get(storefront_url, timeout=15, allow_redirects=True)
    return r.status_code == 200


def build_product_snapshot(sku):
    product = find_product_by_sku(sku)
    if product is None:
        return None
    category_ids = category_ids_of(product)
    snapshot = {
        "status": product["status"],
        "visibility": visibility_of(product),
        "websiteIds": product_website_ids(sku),
        "categoryIds": category_ids,
    }
    categories = fetch_categories(category_ids)
    return snapshot, categories


def diagnose(sku, storefront_url=None):
    built = build_product_snapshot(sku)
    if built is None:
        return {"sku": sku, "status": "not_found"}

    snapshot, categories = built
    verdict = decide_storefront_eligibility(snapshot, categories, TARGET_WEBSITE_ID)

    if not verdict["eligible"]:
        return {"sku": sku, "status": "ineligible", "reasons": verdict["reasons"]}

    if storefront_url and not storefront_has_product(storefront_url):
        return {"sku": sku, "status": "indexer_or_cron_suspected", "suspects": SUSPECT_INDEXERS}

    return {"sku": sku, "status": "ok"}


def repair_product(sku, fixes):
    """fixes may include status, visibility, and/or website_ids (a FULL list).
    Never send a partial website_ids list; that is the documented bug (#11324)
    that reassigns or drops websites. Always prints a diff before writing.
    """
    body = {"product": {"sku": sku}}
    if "status" in fixes:
        body["product"]["status"] = fixes["status"]
    if "visibility" in fixes:
        body["product"]["visibility"] = fixes["visibility"]
    if "website_ids" in fixes:
        body["product"]["extension_attributes"] = {"website_ids": fixes["website_ids"]}

    log.info("DRY_RUN diff for %s: %s", sku, body)
    if DRY_RUN:
        return {"sku": sku, "applied": False, "dry_run": True, "body": body}

    magento_put(f"/products/{sku}", body)
    return {"sku": sku, "applied": True, "dry_run": False, "body": body}


def run(skus, storefront_urls=None):
    storefront_urls = storefront_urls or {}
    reports = []
    for sku in skus:
        report = diagnose(sku, storefront_urls.get(sku))
        log.info("SKU %s: %s", sku, report["status"])
        reports.append(report)
    log.info("Done. %d SKU(s) checked.", len(reports))
    return reports


if __name__ == "__main__":
    target_skus = [s.strip() for s in os.environ.get("TARGET_SKUS", "").split(",") if s.strip()]
    run(target_skus)

File: enabled-product-missing-from-storefront/node/diagnose-missing-product.js
/**
 * Diagnose a Magento 2 product that shows Enabled but is missing from the storefront.
 *
 * Status Enabled is only one of several conditions Magento checks. Visibility
 * has to include catalog or search, the product has to carry the storefront's
 * website_id in its website assignment (a REST create/update that omits
 * extension_attributes.website_ids can silently drop or fail to set this, per
 * magento2 GitHub issues #8173, #10495, #11324), and it has to link to at
 * least one active category. Even when all three agree, a stale or invalid
 * indexer (catalog_category_product, catalog_product_index,
 * catalogsearch_fulltext) or a missed cron run can still hide the product,
 * and that can only be fixed with bin/magento indexer:reindex, not over REST.
 * This reports by default. Run on a schedule or on demand. Safe to run again
 * and again.
 *
 * Guide: https://www.allanninal.dev/magento/enabled-product-missing-from-storefront/
 */
import { pathToFileURL } from "node:url";

const MAGENTO_URL = (process.env.MAGENTO_URL || "https://demo.example.com").replace(/\/$/, "");
const TOKEN = process.env.MAGENTO_ADMIN_TOKEN || "token_dummy";
const TARGET_WEBSITE_ID = Number(process.env.TARGET_WEBSITE_ID || 1);
const DRY_RUN = (process.env.DRY_RUN || "true").toLowerCase() === "true";

const NOT_VISIBLE_INDIVIDUALLY = 1;
const SUSPECT_INDEXERS = ["catalog_category_product", "catalog_product_index", "catalogsearch_fulltext"];

/**
 * Pure decision logic, no I/O.
 *
 * product: {status: 1|2, visibility: 1|2|3|4, websiteIds: number[], categoryIds: number[]}
 * categories: {id: number, isActive: boolean}[]
 * targetWebsiteId: number
 * Returns {eligible: boolean, reasons: string[]}.
 *
 * eligible is true only if status === 1 AND visibility !== 1 (Not Visible
 * Individually) AND websiteIds includes targetWebsiteId AND at least one
 * linked category is active. Otherwise reasons lists every failing
 * condition, so the caller can tell "should be eligible per data but still
 * missing from storefront" (stale index/cron) apart from "genuinely
 * ineligible per data".
 */
export function decideStorefrontEligibility(product, categories, targetWebsiteId) {
  const reasons = [];

  if (product.status !== 1) reasons.push("disabled");
  if (product.visibility === NOT_VISIBLE_INDIVIDUALLY) reasons.push("not_visible_individually");
  if (!product.websiteIds.includes(targetWebsiteId)) reasons.push("website_not_assigned");

  const activeIds = new Set(categories.filter((c) => c.isActive).map((c) => c.id));
  const hasActiveCategory = product.categoryIds.some((id) => activeIds.has(id));
  if (!hasActiveCategory) reasons.push("no_active_category");

  return { eligible: reasons.length === 0, reasons };
}

async function magentoGet(path, params = {}) {
  const url = new URL(`${MAGENTO_URL}/rest/V1${path}`);
  for (const [k, v] of Object.entries(params)) url.searchParams.set(k, v);
  const res = await fetch(url, { headers: { Authorization: `Bearer ${TOKEN}` } });
  if (!res.ok) throw new Error(`Magento ${res.status}`);
  return res.json();
}

async function magentoPut(path, body) {
  const res = await fetch(`${MAGENTO_URL}/rest/V1${path}`, {
    method: "PUT",
    headers: { Authorization: `Bearer ${TOKEN}`, "Content-Type": "application/json" },
    body: JSON.stringify(body),
  });
  if (!res.ok) throw new Error(`Magento ${res.status}`);
  return res.json();
}

async function findProductBySku(sku) {
  const params = {
    "searchCriteria[filterGroups][0][filters][0][field]": "sku",
    "searchCriteria[filterGroups][0][filters][0][value]": sku,
    "searchCriteria[filterGroups][0][filters][0][conditionType]": "eq",
  };
  const data = await magentoGet("/products", params);
  return data.items[0] || null;
}

function visibilityOf(product) {
  const attr = (product.custom_attributes || []).find((a) => a.attribute_code === "visibility");
  return attr ? Number(attr.value) : null;
}

function categoryIdsOf(product) {
  const attr = (product.custom_attributes || []).find((a) => a.attribute_code === "category_ids");
  return attr ? attr.value.map(Number) : [];
}

async function productWebsiteIds(sku) {
  return magentoGet(`/products/${sku}/websites`);
}

async function fetchCategories(categoryIds) {
  const result = [];
  for (const id of categoryIds) {
    const data = await magentoGet(`/categories/${id}`);
    result.push({ id, isActive: Boolean(data.is_active) });
  }
  return result;
}

async function storefrontHasProduct(storefrontUrl) {
  const res = await fetch(storefrontUrl, { redirect: "follow" });
  return res.status === 200;
}

async function buildProductSnapshot(sku) {
  const product = await findProductBySku(sku);
  if (!product) return null;
  const categoryIds = categoryIdsOf(product);
  const snapshot = {
    status: product.status,
    visibility: visibilityOf(product),
    websiteIds: await productWebsiteIds(sku),
    categoryIds,
  };
  const categories = await fetchCategories(categoryIds);
  return { snapshot, categories };
}

export async function diagnose(sku, storefrontUrl) {
  const built = await buildProductSnapshot(sku);
  if (!built) return { sku, status: "not_found" };

  const { snapshot, categories } = built;
  const verdict = decideStorefrontEligibility(snapshot, categories, TARGET_WEBSITE_ID);

  if (!verdict.eligible) {
    return { sku, status: "ineligible", reasons: verdict.reasons };
  }

  if (storefrontUrl && !(await storefrontHasProduct(storefrontUrl))) {
    return { sku, status: "indexer_or_cron_suspected", suspects: SUSPECT_INDEXERS };
  }

  return { sku, status: "ok" };
}

export async function repairProduct(sku, fixes) {
  // fixes may include status, visibility, and/or websiteIds (a FULL list).
  // Never send a partial websiteIds list; that is the documented bug (#11324)
  // that reassigns or drops websites. Always prints a diff before writing.
  const body = { product: { sku } };
  if ("status" in fixes) body.product.status = fixes.status;
  if ("visibility" in fixes) body.product.visibility = fixes.visibility;
  if ("websiteIds" in fixes) body.product.extension_attributes = { website_ids: fixes.websiteIds };

  console.log(`DRY_RUN diff for ${sku}:`, JSON.stringify(body));
  if (DRY_RUN) return { sku, applied: false, dryRun: true, body };

  await magentoPut(`/products/${sku}`, body);
  return { sku, applied: true, dryRun: false, body };
}

export async function run(skus, storefrontUrls = {}) {
  const reports = [];
  for (const sku of skus) {
    const report = await diagnose(sku, storefrontUrls[sku]);
    console.log(`SKU ${sku}: ${report.status}`);
    reports.push(report);
  }
  console.log(`Done. ${reports.length} SKU(s) checked.`);
  return reports;
}

if (import.meta.url === pathToFileURL(process.argv[1]).href) {
  const targetSkus = (process.env.TARGET_SKUS || "").split(",").map((s) => s.trim()).filter(Boolean);
  run(targetSkus).catch((err) => { console.error(err); process.exit(1); });
}
